eclipse-theia/theia
# Encapsulate Electron Window Logic - Allow Exit during Load

#### What it does

Fixes #10586
This PR is a pretty big intervention into the logic of the ElectronMainApplication. The narrow goal is to track whether a window on the frontend is in a state to handle a close request so that windows that aren't ready to handle the close request can be closed immediately. The larger goal is to begin to extract the logic of handling Theia windows so that we can develop an abstraction that will work for both Electron and browser with an eye to doing things like running cleanup when the last window is closed. There are much smaller interventions that can address the narrow goal, so if there isn't an appetite for this size of refactor, that's fine, and I'm happy to reduce it to the minimum necessary to ensure that close gets handled correctly.

#### How to test

> It would be nice to write automated tests for some of this logic, but I haven't interacted with our Electron-specific test system much. Any pointers are appreciated.
1. Add some code that will prevent the frontend from ever achieving a `ready` state.
> e.g. add a `return` to `FrontendApplication#start` that terminates it early.
2. Run the application and try to close the window.
3. It should close.

---
1. Let the application get ready normally.
2. Dirty a file, or do something else that would generate a close prompt.
3. Observe that the prompt is displayed correctly and the application closes or doesn't close correctly
4. Change the `window.titleBarStyle` (if not on Mac) preference and observe that the application restarts correctly.


#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
